<a href="https://colab.research.google.com/github/harakenken22/diveintocode-ml/blob/master/Graduation_production.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
import pandas as pd
import numpy as np
from keras.datasets import mnist
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import f1_score
from sklearn.metrics import confusion_matrix
import math
import matplotlib.pyplot as plt
%matplotlib inline

df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/博士研究/All_Data_Kiambu_Exploratory_200422_treated_data.csv')

In [4]:
#データの確認
df

start time  ... BMI_prediction_result
0    2019-11-04T09:49:23+0000  ...                 FALSE
1    2019-10-28T07:39:16+0000  ...                 FALSE
2    2019-10-28T07:18:08+0000  ...                 FALSE
3    2019-10-28T07:58:15+0000  ...                 FALSE
4    2019-10-21T05:58:33+0000  ...                 FALSE
..                        ...  ...                   ...
127  2019-10-07T05:41:36+0000  ...                 FALSE
128  2019-10-28T07:58:15+0000  ...                 FALSE
129  2019-10-07T06:17:02+0000  ...                  TURE
130  2019-11-11T06:39:32+0000  ...                  TURE
131  2019-09-23T06:21:35+0000  ...                  TURE

[132 rows x 219 columns]

In [5]:
df.describe()

Household ID number  ...  BMI_EVENT_Prediction
count           132.000000  ...            132.000000
mean              4.242424  ...              3.053030
std               3.426705  ...              0.583676
min               1.000000  ...              2.000000
25%               2.000000  ...              3.000000
50%               3.000000  ...              3.000000
75%               5.000000  ...              3.000000
max              16.000000  ...              4.000000

[8 rows x 205 columns]

In [6]:
#forでdescribeを行う場合
for i in range(len(df.columns)):
  print(df.iloc[:,i].describe())

count                          132
unique                          72
top       2019-09-23T11:09:53+0000
freq                             2
Name: start time, dtype: object
count         132
unique         23
top       Muthumu
freq           16
Name: Village Name, dtype: object
count                      132
unique                      15
top       Ruku shopping center
freq                        14
Name: Household ID, dtype: object
count    132.000000
mean       4.242424
std        3.426705
min        1.000000
25%        2.000000
50%        3.000000
75%        5.000000
max       16.000000
Name: Household ID number, dtype: float64
count    132.000000
mean       4.553030
std        1.394178
min        2.000000
25%        4.000000
50%        5.000000
75%        5.000000
max       10.000000
Name: Number of people sharing the same pot, dtype: float64
count              132
unique             131
top       David Mungai
freq                 2
Name: Husband full name, dtype: object
count    13

In [7]:
#特徴量をすべて出す
for i in range(len(df.columns)):
  print('{},'.format(df.columns[i]))

start time,
Village Name,
Household ID,
Household ID number,
Number of people sharing the same pot,
Husband full name,
Age,
Gender,
Education level of husband,
Major occupation of Husband,
Fulltime farmer,
Religious leader,
Casual laborer,
Civil servant,
Business person,
Student,
Other (specify),
Religion of Husband,
Mobile phone,
Bicycle,
Motorbike,
Car/truck,
Radio / stereo,
TV set or DVD,
Satellite dish,
Refrigerator,
Own stand pipe,
Own borehole/well,
Own water tank,
Access to shared well/borehole/stand pipe,
Latrine/toilet,
Other:,
Wood,
Charcoal,
Biogas (stove),
Electricity,
Solar panel,
Battery (large, e.g. car battery for power),
Other:__1,
Gas,
Monthly income(Ksh),
MIL,
Gender__1,
Number of cultivation field,
walking min from your house to Wangige market?,
walking min from your house to your favorite market,
African nightshade_V,
Amaranth_V,
Apple local_F,
Avocado_F,
Banana_F,
Beans_P,
Beetroot_R,
Broccoli_V,
Cabbage_V,
Carrot_V,
Cassava_R,
Chili pepper_S,
Coco-yam_R,
Coriande

In [8]:
# #変数に入れてilocで任意の値を取得
# Desc.iloc[:, 0]

In [9]:
# for i in range(len(df.columns)):
#   df[i].dtypes

In [10]:
#for文でデータの確認
for i in range(len(df.columns)):
  print(df.iloc[:,i].dtypes)

object
object
object
int64
int64
object
int64
object
object
object
int64
int64
int64
int64
int64
int64
int64
object
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
float64
object
float64
object
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
object
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
i

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132 entries, 0 to 131
Columns: 219 entries, start time to BMI_prediction_result
dtypes: float64(32), int64(173), object(14)
memory usage: 226.0+ KB


In [12]:
df.value_counts

<bound method DataFrame.value_counts of                    start time  ... BMI_prediction_result
0    2019-11-04T09:49:23+0000  ...                 FALSE
1    2019-10-28T07:39:16+0000  ...                 FALSE
2    2019-10-28T07:18:08+0000  ...                 FALSE
3    2019-10-28T07:58:15+0000  ...                 FALSE
4    2019-10-21T05:58:33+0000  ...                 FALSE
..                        ...  ...                   ...
127  2019-10-07T05:41:36+0000  ...                 FALSE
128  2019-10-28T07:58:15+0000  ...                 FALSE
129  2019-10-07T06:17:02+0000  ...                  TURE
130  2019-11-11T06:39:32+0000  ...                  TURE
131  2019-09-23T06:21:35+0000  ...                  TURE

[132 rows x 219 columns]>

In [13]:
df.isnull()

start time  Village Name  ...  BMI_EVENT_Prediction  BMI_prediction_result
0         False         False  ...                 False                  False
1         False         False  ...                 False                  False
2         False         False  ...                 False                  False
3         False         False  ...                 False                  False
4         False         False  ...                 False                  False
..          ...           ...  ...                   ...                    ...
127       False         False  ...                 False                  False
128       False         False  ...                 False                  False
129       False         False  ...                 False                  False
130       False         False  ...                 False                  False
131       False         False  ...                 False                  False

[132 rows x 219 columns]

特徴量の選択：
Number of people sharing the same pot,
Age,
Gender,
Education level of husband,
MIL,
Gender__1,
Number of cultivation field,
walking min from your house to Wangige market?,
Cultivation diversity,
Cultivation diversity_Cereals,
Cultivation diversity_Root,
Cultivation diversity_Suger,
Cultivation diversity_Vegetable,
Cultivation diversity_Palse,
Cultivation diversity_Fruits,
Cultivation diversity_Spice,
Wild Diversity,
Wild Diversity_Cereals,
Wild Diversity_Root,
Wild Diversity_Suger,
Wild Diversity_Vegetable,
Wild Diversity_Palse,
Wild Diversity_Fruits,
Wild Diversity_Spice,
Animal Diversity,
Agrobiodiversity,
Height(cm),
Weight(kg),
BMI,
Waist(cm),
Hip(cm),
Ratio W/H,
Body Fat (%),
Systolic blood pressure(mmHg),
Diastolic blood pressure(mmHg),
Event_hypertension,
Grasping power(kg),
Total_happness,
Activity_level_2,
Steps,
percentage_production,
Sleeping_time,
Economics_time,
Crop_time,
Livestock_time,
BMI_EVENT,
Gender_replaced,

In [14]:
#Statsmodelを用いて

参考文献：https://tanuhack.com/statsmodels-multiple-lra/

In [15]:
df_features = df[['Household ID',
'Household ID number',
'Number of people sharing the same pot',
'Age',
'Gender',
'Education level of husband',
'MIL',
'Number of cultivation field',
'walking min from your house to Wangige market?',
'Cultivation diversity',
'Cultivation diversity_Cereals',
'Cultivation diversity_Root',
'Cultivation diversity_Suger',
'Cultivation diversity_Vegetable',
'Cultivation diversity_Palse',
'Cultivation diversity_Fruits',
'Cultivation diversity_Spice',
'Wild Diversity',
'Wild Diversity_Cereals',
'Wild Diversity_Root',
'Wild Diversity_Suger',
'Wild Diversity_Vegetable',
'Wild Diversity_Palse',
'Wild Diversity_Fruits',
'Wild Diversity_Spice',
'Animal Diversity',
'Agrobiodiversity',
'Height(cm)',
'Weight(kg)',
'BMI',
'Waist(cm)',
'Hip(cm)',
'Ratio W/H',
'Body Fat (%)',
'Systolic blood pressure(mmHg)',
'Diastolic blood pressure(mmHg)',
'Event_hypertension',
'Grasping power(kg)',
'Total_happness',
'Activity_level_2',
'Steps',
'percentage_production',
'Sleeping_time',
'Economics_time',
'Crop_time',
'Livestock_time',
'BMI_EVENT',
'Gender_replaced']]

In [16]:
df_features

Household ID  Household ID number  ...  BMI_EVENT  Gender_replaced
0    Muthumu/ Kangora                    2  ...          3                1
1             Muthumu                    3  ...          2                0
2             Muthumu                    2  ...          4                0
3             Muthumu                    4  ...          4                0
4              Gababa                    5  ...          2                0
..                ...                  ...  ...        ...              ...
127       Gatathuthua                    5  ...          4                0
128           Muthumu                    4  ...          3                1
129           Kangora                    1  ...          2                1
130           Wangige                    2  ...          2                1
131           Kiiniki                    3  ...          3                0

[132 rows x 48 columns]

In [17]:
df_time_consuption = df.loc[:, ['Household ID',
'Household ID number',
'Sleeping_time',
'Economics_time',
'Crop_time',
'Livestock_time']]

In [18]:
df_time_consuption

Household ID  Household ID number  ...   Crop_time  Livestock_time
0    Muthumu/ Kangora                    2  ...  390.000000       90.000000
1             Muthumu                    3  ...  270.000000      120.000000
2             Muthumu                    2  ...  210.000000        0.000000
3             Muthumu                    4  ...  272.210526       49.263158
4              Gababa                    5  ...  480.000000        0.000000
..                ...                  ...  ...         ...             ...
127       Gatathuthua                    5  ...  272.210526       49.263158
128           Muthumu                    4  ...  272.210526       49.263158
129           Kangora                    1  ...   90.000000      120.000000
130           Wangige                    2  ...    0.000000        0.000000
131           Kiiniki                    3  ...  272.210526       49.263158

[132 rows x 6 columns]

StatsModelsモジュールの使い方
----

In [19]:
import pandas as pd
import statsmodels.api as sm
import statsmodels.formula.api as smf

/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [20]:
!pip install statsmodels

In [21]:
# get_dummies()で質的データも対応可能

x = df[[
'Number of people sharing the same pot',
'Age',
'MIL',
'Number of cultivation field',
'walking min from your house to Wangige market?',
'Cultivation diversity',
'Cultivation diversity_Cereals',
'Cultivation diversity_Root',
'Cultivation diversity_Suger',
'Cultivation diversity_Vegetable',
'Cultivation diversity_Palse',
'Cultivation diversity_Fruits',
'Cultivation diversity_Spice',
'Wild Diversity',
'Wild Diversity_Cereals',
'Wild Diversity_Root',
'Wild Diversity_Suger',
'Wild Diversity_Vegetable',
'Wild Diversity_Palse',
'Wild Diversity_Fruits',
'Wild Diversity_Spice',
'Animal Diversity',
'Agrobiodiversity',
'Height(cm)',
'Weight(kg)',
'BMI',
'Waist(cm)',
'Hip(cm)',
'Ratio W/H',
'Body Fat (%)',
'Systolic blood pressure(mmHg)',
'Diastolic blood pressure(mmHg)',
'Grasping power(kg)',
'Total_happness',
'Activity_level_2',
'Steps',
'percentage_production',
'Sleeping_time',
'Economics_time',
'Crop_time',
'Livestock_time',
'Gender_replaced']]


y = df['BMI_EVENT']

In [22]:
x
x.dtypes

Number of people sharing the same pot               int64
Age                                                 int64
MIL                                               float64
Number of cultivation field                         int64
walking min from your house to Wangige market?      int64
Cultivation diversity                               int64
Cultivation diversity_Cereals                       int64
Cultivation diversity_Root                          int64
Cultivation diversity_Suger                         int64
Cultivation diversity_Vegetable                     int64
Cultivation diversity_Palse                         int64
Cultivation diversity_Fruits                        int64
Cultivation diversity_Spice                         int64
Wild Diversity                                      int64
Wild Diversity_Cereals                              int64
Wild Diversity_Root                                 int64
Wild Diversity_Suger                                int64
Wild Diversity

In [23]:
y

0      3
1      2
2      4
3      4
4      2
      ..
127    4
128    3
129    2
130    2
131    3
Name: BMI_EVENT, Length: 132, dtype: int64

In [24]:
x_nparray = np.array(x)

In [25]:
y_nparray = np.array(y)

In [26]:
# 定数項(y切片)を必要とする線形回帰のモデル式ならば必須
X = sm.add_constant(x.astype(float))
print(X)

# 最小二乗法でモデル化
model = sm.OLS(y.astype(float), X)
result = model.fit()

# 重回帰分析の結果を表示
result.summary()

     const  ...  Gender_replaced
0      1.0  ...              1.0
1      1.0  ...              0.0
2      1.0  ...              0.0
3      1.0  ...              0.0
4      1.0  ...              0.0
..     ...  ...              ...
127    1.0  ...              0.0
128    1.0  ...              1.0
129    1.0  ...              1.0
130    1.0  ...              1.0
131    1.0  ...              0.0

[132 rows x 43 columns]


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              BMI_EVENT   R-squared:                       0.684
Model:                            OLS   Adj. R-squared:                  0.559
Method:                 Least Squares   F-statistic:                     5.492
Date:                Wed, 05 May 2021   Prob (F-statistic):           1.30e-11
Time:                        04:21:38   Log-Likelihood:                -101.49
No. Observations:                 132   AIC:                             279.0
Df Residuals:                      94   BIC:                             388.5
Df Model:                          37                                         
Covariance Type:            nonrobust                                         
==================================================================================================================
                                                     coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------------------
const                                             -4.2554     10.046     -0.424      0.673     -24.202      15.691
Number of people sharing the same pot             -0.0088      0.050     -0.176      0.861      -0.108       0.091
Age                                               -0.0075      0.008     -0.892      0.374      -0.024       0.009
MIL                                                0.0081      0.033      0.245      0.807      -0.058       0.074
Number of cultivation field                        0.1102      0.116      0.953      0.343      -0.119       0.340
walking min from your house to Wangige market?    -0.0052      0.008     -0.668      0.506      -0.021       0.010
Cultivation diversity                              0.1877      0.452      0.415      0.679      -0.710       1.086
Cultivation diversity_Cereals                     -0.1093      0.655     -0.167      0.868      -1.409       1.190
Cultivation diversity_Root                        -0.2354      0.624     -0.377      0.707      -1.475       1.004
Cultivation diversity_Suger                       -0.2906      0.585     -0.497      0.621      -1.452       0.871
Cultivation diversity_Vegetable                   -0.2829      0.606     -0.467      0.642      -1.486       0.921
Cultivation diversity_Palse                       -0.2872      0.623     -0.461      0.646      -1.523       0.949
Cultivation diversity_Fruits                      -0.0725      0.625     -0.116      0.908      -1.314       1.169
Cultivation diversity_Spice                       -0.1756      0.616     -0.285      0.776      -1.399       1.047
Wild Diversity                                    -0.0832      0.166     -0.503      0.616      -0.412       0.245
Wild Diversity_Cereals                          7.704e-14   5.73e-14      1.344      0.182   -3.68e-14    1.91e-13
Wild Diversity_Root                             1.272e-14   1.27e-14      1.001      0.319   -1.25e-14    3.79e-14
Wild Diversity_Suger                               0.1618      0.234      0.692      0.490      -0.302       0.626
Wild Diversity_Vegetable                           0.0749      0.201      0.373      0.710      -0.323       0.473
Wild Diversity_Palse                           -5.302e-14   4.41e-14     -1.202      0.232   -1.41e-13    3.46e-14
Wild Diversity_Fruits                             -0.0978      0.145     -0.674      0.502      -0.386       0.190
Wild Diversity_Spice                              -0.2221      0.439     -0.505      0.614      -1.095       0.651
Animal Diversity                                  -0.0535      0.159     -0.335      0.738      -0.370       0.263
Agrobiodiversity                                   0.0509      0.162      0.315      0.753   

**多重共線性の確認：**
---
Cond. No.	1.17e+16が非常に大きいため、多重共線性があるため、多重共線性をさらに確認する。

In [27]:
from statsmodels.stats.outliers_influence import variance_inflation_factor as vif

num_cols = model.exog.shape[1] # 説明変数の列数
vifs = [vif(model.exog, i) for i in range(0, num_cols)]

pd.DataFrame(vifs, index=model.exog_names, columns=['VIF'])

/usr/local/lib/python3.7/dist-packages/statsmodels/stats/outliers_influence.py:185: RuntimeWarning: divide by zero encountered in double_scalars
  vif = 1. / (1. - r_squared_i)
/usr/local/lib/python3.7/dist-packages/statsmodels/regression/linear_model.py:1636: RuntimeWarning: invalid value encountered in double_scalars
  return 1 - self.ssr/self.centered_tss


VIF
const                                           3.481131e+04
Number of people sharing the same pot           1.675164e+00
Age                                             2.296600e+00
MIL                                             1.478313e+00
Number of cultivation field                     1.636459e+00
walking min from your house to Wangige market?  2.337493e+00
Cultivation diversity                                    inf
Cultivation diversity_Cereals                   4.074967e+01
Cultivation diversity_Root                      3.799976e+01
Cultivation diversity_Suger                     9.022607e+00
Cultivation diversity_Vegetable                 5.962198e+02
Cultivation diversity_Palse                     3.305624e+01
Cultivation diversity_Fruits                    5.674020e+01
Cultivation diversity_Spice                     4.711458e+01
Wild Diversity                                           inf
Wild Diversity_Cereals                                   NaN
Wild Diversity_Root                                      NaN
Wild Diversity_Suger                                     inf
Wild Diversity_Vegetable                                 inf
Wild Diversity_Palse                                     NaN
Wild Diversity_Fruits                                    inf
Wild Diversity_Spice                                     inf
Animal Diversity                                         inf
Agrobiodiversity                                         inf
Height(cm)                                      4.956037e+01
Weight(kg)                                      1.681524e+02
BMI                                             1.833759e+02
Waist(cm)                                       3.745017e+02
Hip(cm)                                         1.585649e+02
Ratio W/H                                       1.946005e+02
Body Fat (%)                                    2.520075e+00
Systolic blood pressure(mmHg)                   2.607116e+00
Diastolic blood pressure(mmHg)                  3.369160e+00
Grasping power(kg)                              2.887395e+00
Total_happness                                  1.347525e+00
Activity_level_2                                3.247359e+00
Steps                                           2.786111e+00
percentage_production                           1.614925e+00
Sleeping_time                                   1.432214e+00
Economics_time                                  2.436279e+00
Crop_time                                       2.847623e+00
Livestock_time                                  1.874280e+00
Gender_replaced                                 3.346345e+00

データセットの再検討：
下記のように１０以上の値をとっているものを再検討を行う
- Wild Diversity	inf
- Wild Diversity_Cereals	NaN
- Wild Diversity_Root	NaN
- Wild Diversity_Suger	inf
- Wild Diversity_Vegetable	inf
- Wild Diversity_Palse	NaN
- Wild Diversity_Fruits	inf
- Wild Diversity_Spice	inf
- Animal Diversity	inf
- Agrobiodiversity	inf
- Cultivation diversity	inf

説明変数をWild Diversity, Animal, CultivationをAgrobiodiversityに統合


In [28]:
x_removed = df[[
'Number of people sharing the same pot',
'Age',
'MIL',
'Number of cultivation field',
'walking min from your house to Wangige market?',
'Agrobiodiversity',
'Height(cm)',
'Weight(kg)',
'BMI',
'Waist(cm)',
'Hip(cm)',
'Ratio W/H',
'Body Fat (%)',
'Systolic blood pressure(mmHg)',
'Diastolic blood pressure(mmHg)',
'Grasping power(kg)',
'Total_happness',
'Activity_level_2',
'Steps',
'percentage_production',
'Sleeping_time',
'Economics_time',
'Crop_time',
'Livestock_time',
'Gender_replaced']]


y = df['BMI_EVENT']

In [29]:
#２回目の検定
# 定数項(y切片)を必要とする線形回帰のモデル式ならば必須
X = sm.add_constant(x_removed.astype(float))
print(X)

# 最小二乗法でモデル化
model_2nd = sm.OLS(y.astype(float), X)
result_2nd = model_2nd.fit()

# 重回帰分析の結果を表示
result_2nd.summary()

     const  ...  Gender_replaced
0      1.0  ...              1.0
1      1.0  ...              0.0
2      1.0  ...              0.0
3      1.0  ...              0.0
4      1.0  ...              0.0
..     ...  ...              ...
127    1.0  ...              0.0
128    1.0  ...              1.0
129    1.0  ...              1.0
130    1.0  ...              1.0
131    1.0  ...              0.0

[132 rows x 26 columns]


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              BMI_EVENT   R-squared:                       0.658
Model:                            OLS   Adj. R-squared:                  0.577
Method:                 Least Squares   F-statistic:                     8.154
Date:                Wed, 05 May 2021   Prob (F-statistic):           3.25e-15
Time:                        04:21:38   Log-Likelihood:                -106.67
No. Observations:                 132   AIC:                             265.3
Df Residuals:                     106   BIC:                             340.3
Df Model:                          25                                         
Covariance Type:            nonrobust                                         
==================================================================================================================
                                                     coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------------------
const                                              0.4511      9.548      0.047      0.962     -18.479      19.382
Number of people sharing the same pot              0.0114      0.045      0.256      0.798      -0.077       0.100
Age                                               -0.0058      0.007     -0.823      0.412      -0.020       0.008
MIL                                                0.0019      0.029      0.066      0.947      -0.056       0.060
Number of cultivation field                        0.0913      0.102      0.897      0.372      -0.111       0.293
walking min from your house to Wangige market?    -0.0002      0.006     -0.028      0.978      -0.011       0.011
Agrobiodiversity                                  -0.0071      0.014     -0.525      0.601      -0.034       0.020
Height(cm)                                        -0.0442      0.041     -1.080      0.283      -0.125       0.037
Weight(kg)                                         0.0492      0.048      1.029      0.306      -0.046       0.144
BMI                                               -0.0015      0.129     -0.011      0.991      -0.257       0.254
Waist(cm)                                         -0.0650      0.063     -1.039      0.301      -0.189       0.059
Hip(cm)                                            0.0605      0.056      1.089      0.279      -0.050       0.171
Ratio W/H                                          6.6223      6.417      1.032      0.304      -6.099      19.344
Body Fat (%)                                       0.0024      0.008      0.302      0.763      -0.014       0.019
Systolic blood pressure(mmHg)                      0.0027      0.005      0.582      0.562      -0.007       0.012
Diastolic blood pressure(mmHg)                     0.0010      0.007      0.139      0.890      -0.014       0.016
Grasping power(kg)                                 0.0067      0.008      0.819      0.415      -0.010       0.023
Total_happness                                    -0.0234      0.019     -1.251      0.214      -0.061       0.014
Activity_level_2                                   0.0009      0.001      0.818      0.415      -0.001       0.003
Steps                                          -9.461e-06   1.02e-05     -0.932      0.354   -2.96e-05    1.07e-05
percentage_production                             -0.0020      0.004     -0.482      0.630      -0.010       0.006
Sleeping_time                                  -9.894e-05      0.001     -0.144      0.885      -0.001       0.001
Economics_time                                     0.0010      0.000      2.172      0.032    8.96e-05       0.002
Crop_time                                         -0.0008      0.000     -1.644      0.103   

In [30]:
from statsmodels.stats.outliers_influence import variance_inflation_factor as vif

num_cols_2nd = model_2nd.exog.shape[1] # 説明変数の列数
vifs = [vif(model_2nd.exog, i) for i in range(0, num_cols_2nd)]

pd.DataFrame(vifs, index=model_2nd.exog_names, columns=['VIF'])

VIF
const                                           32786.587603
Number of people sharing the same pot               1.386860
Age                                                 1.668290
MIL                                                 1.190995
Number of cultivation field                         1.321549
walking min from your house to Wangige market?      1.283180
Agrobiodiversity                                    1.258090
Height(cm)                                         45.530315
Weight(kg)                                        156.046045
BMI                                               169.437731
Waist(cm)                                         341.984847
Hip(cm)                                           144.916837
Ratio W/H                                         177.860851
Body Fat (%)                                        2.202328
Systolic blood pressure(mmHg)                       2.382479
Diastolic blood pressure(mmHg)                      2.761494
Grasping power(kg)                                  2.449838
Total_happness                                      1.260358
Activity_level_2                                    2.736562
Steps                                               2.526943
percentage_production                               1.279874
Sleeping_time                                       1.270390
Economics_time                                      2.063991
Crop_time                                           2.435373
Livestock_time                                      1.716771
Gender_replaced                                     2.891690

データセットの再検討： 下記のように１０以上の値をとっているものを再検討を行う

- Height(cm)	45.530315
- Weight(kg)	156.046045
- BMI	169.437731
- Waist(cm)	341.984847
- Hip(cm)	144.916837
- Ratio W/H	177.860851

上記もVIFが高かっため、説明変数から削除する。
また、身体的特徴に関する項目も削除した。

In [31]:
x_feature = df[[
'Number of people sharing the same pot',
'Age',
'MIL',
'Number of cultivation field',
'walking min from your house to Wangige market?',
'Agrobiodiversity',
'Grasping power(kg)',
'Total_happness',
'Activity_level_2',
'Steps',
'percentage_production',
'Sleeping_time',
'Economics_time',
'Crop_time',
'Livestock_time',
'Gender_replaced']]


y_objective = df['BMI_EVENT']

In [32]:
#３回目の検定
# 定数項(y切片)を必要とする線形回帰のモデル式ならば必須
X = sm.add_constant(x_feature.astype(float))
print(X)

# 最小二乗法でモデル化
model_3rd = sm.OLS(y_objective.astype(float), X)
result_3rd = model_3rd.fit()

# 重回帰分析の結果を表示
result_3rd.summary()

     const  ...  Gender_replaced
0      1.0  ...              1.0
1      1.0  ...              0.0
2      1.0  ...              0.0
3      1.0  ...              0.0
4      1.0  ...              0.0
..     ...  ...              ...
127    1.0  ...              0.0
128    1.0  ...              1.0
129    1.0  ...              1.0
130    1.0  ...              1.0
131    1.0  ...              0.0

[132 rows x 17 columns]


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              BMI_EVENT   R-squared:                       0.092
Model:                            OLS   Adj. R-squared:                 -0.034
Method:                 Least Squares   F-statistic:                    0.7326
Date:                Wed, 05 May 2021   Prob (F-statistic):              0.756
Time:                        04:21:38   Log-Likelihood:                -171.06
No. Observations:                 132   AIC:                             376.1
Df Residuals:                     115   BIC:                             425.1
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
==================================================================================================================
                                                     coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------------------
const                                              2.1233      1.189      1.785      0.077      -0.232       4.479
Number of people sharing the same pot              0.0447      0.063      0.707      0.481      -0.080       0.170
Age                                                0.0010      0.010      0.104      0.918      -0.018       0.020
MIL                                                0.0760      0.043      1.748      0.083      -0.010       0.162
Number of cultivation field                        0.1229      0.151      0.816      0.416      -0.175       0.421
walking min from your house to Wangige market?     0.0041      0.008      0.487      0.627      -0.013       0.021
Agrobiodiversity                                  -0.0145      0.020     -0.710      0.479      -0.055       0.026
Grasping power(kg)                                 0.0124      0.011      1.086      0.280      -0.010       0.035
Total_happness                                    -0.0216      0.028     -0.772      0.442      -0.077       0.034
Activity_level_2                                  -0.0015      0.002     -0.916      0.361      -0.005       0.002
Steps                                           1.616e-05   1.52e-05      1.061      0.291    -1.4e-05    4.63e-05
percentage_production                              0.0004      0.006      0.063      0.950      -0.012       0.012
Sleeping_time                                      0.0002      0.001      0.183      0.855      -0.002       0.002
Economics_time                                     0.0011      0.001      1.574      0.118      -0.000       0.002
Crop_time                                         -0.0010      0.001     -1.346      0.181      -0.002       0.000
Livestock_time                                    -0.0006      0.001     -0.448      0.655      -0.003       0.002
Gender_replaced                                   -0.3395      0.241     -1.411      0.161      -0.816       0.137
==============================================================================
Omnibus:                        9.956   Durbin-Watson:                   2.146
Prob(Omnibus):                  0.007   Jarque-Bera (JB):                5.659
Skew:                          -0.326   Prob(JB):                       0.0590
Kurtosis:                       2.223   Cond. No.                     2.32e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.32e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [33]:
from statsmodels.stats.outliers_influence import variance_inflation_factor as vif

num_cols_3rd = model_3rd.exog.shape[1] # 説明変数の列数
vifs = [vif(model_3rd.exog, i) for i in range(0, num_cols_3rd)]

pd.DataFrame(vifs, index=model_3rd.exog_names, columns=['VIF'])

VIF
const                                           208.022104
Number of people sharing the same pot             1.132704
Age                                               1.274538
MIL                                               1.075464
Number of cultivation field                       1.181842
walking min from your house to Wangige market?    1.167275
Agrobiodiversity                                  1.170278
Grasping power(kg)                                1.960943
Total_happness                                    1.148506
Activity_level_2                                  2.579548
Steps                                             2.324161
percentage_production                             1.083411
Sleeping_time                                     1.194405
Economics_time                                    1.773320
Crop_time                                         2.049048
Livestock_time                                    1.594698
Gender_replaced                                   2.123173

機械学習モデルを用いて分析
---
- LIGHT GBM
- ランダムフォレスト
- 重回帰分析
を用いて分析

Light GBMにて分析を行う
---

In [34]:
# データ処理ラブラり
import numpy as np
import pandas as pd
 
# データ可視化ライブラリ
import matplotlib.pyplot as plt
%matplotlib inline 
 
 
# LightGBM
import lightgbm as lgb
 
# Scikit-learn（評価算出）
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score

In [35]:
X_train, X_test, y_train, y_test = train_test_split(x_feature, y_objective, test_size=0.2, random_state=None)

In [36]:
print(X_train)
print(y_train)

     Number of people sharing the same pot  ...  Gender_replaced
41                                       7  ...                1
45                                       4  ...                1
35                                       3  ...                0
44                                       5  ...                1
121                                      5  ...                1
..                                     ...  ...              ...
70                                       4  ...                1
99                                       4  ...                0
102                                      4  ...                0
14                                       4  ...                1
26                                       5  ...                1

[105 rows x 16 columns]
41     4
45     2
35     1
44     4
121    4
      ..
70     2
99     3
102    4
14     2
26     4
Name: BMI_EVENT, Length: 105, dtype: int64


In [37]:
lgb.LGBMClassifier(objective='binary')
# lgb.LGBMClassifier()

LGBMClassifier(boosting_type='gbdt', class_weight=None, colsample_bytree=1.0,
               importance_type='split', learning_rate=0.1, max_depth=-1,
               min_child_samples=20, min_child_weight=0.001, min_split_gain=0.0,
               n_estimators=100, n_jobs=-1, num_leaves=31, objective='binary',
               random_state=None, reg_alpha=0.0, reg_lambda=0.0, silent=True,
               subsample=1.0, subsample_for_bin=200000, subsample_freq=0)

In [38]:
params = {
    'task': 'train',
    'boosting_type': 'gbdt',
    'objective': 'multiclass',
    'num_class': 5,
    'verbose': 4,
}

In [39]:
# 訓練・テストデータの設定
train_data = lgb.Dataset(X_train, label=y_train)
eval_data = lgb.Dataset(X_test, label=y_test, reference= train_data)

In [40]:
gbm = lgb.train(
params,
train_data,
valid_sets=eval_data,
num_boost_round=1000,
verbose_eval=4,
)

[4]	valid_0's multi_logloss: 1.16676
[8]	valid_0's multi_logloss: 1.18872
[12]	valid_0's multi_logloss: 1.18787
[16]	valid_0's multi_logloss: 1.20837
[20]	valid_0's multi_logloss: 1.21693
[24]	valid_0's multi_logloss: 1.22655
[28]	valid_0's multi_logloss: 1.24497
[32]	valid_0's multi_logloss: 1.27149
[36]	valid_0's multi_logloss: 1.27372
[40]	valid_0's multi_logloss: 1.28113
[44]	valid_0's multi_logloss: 1.29777
[48]	valid_0's multi_logloss: 1.31155
[52]	valid_0's multi_logloss: 1.33834
[56]	valid_0's multi_logloss: 1.36
[60]	valid_0's multi_logloss: 1.37377
[64]	valid_0's multi_logloss: 1.387
[68]	valid_0's multi_logloss: 1.40241
[72]	valid_0's multi_logloss: 1.41178
[76]	valid_0's multi_logloss: 1.43448
[80]	valid_0's multi_logloss: 1.44136
[84]	valid_0's multi_logloss: 1.44657
[88]	valid_0's multi_logloss: 1.45976
[92]	valid_0's multi_logloss: 1.48125
[96]	valid_0's multi_logloss: 1.50633
[100]	valid_0's multi_logloss: 1.50717
[104]	valid_0's multi_logloss: 1.50541
[108]	valid_0's m

In [41]:
# y_pred = []
 
# for x in preds:
#     y_pred.append(np.argmax(x))

In [42]:
# confusion_matrix(y_test, y_pred)

In [43]:
# accuracy_score(y_test, y_pred)

Light GBMのグリッドサーチ
---

参考：
https://ailog.site/2021/03/06/2021/0306/

In [44]:
gbm = lgb.LGBMClassifier(objective='binary')
gbm.get_params()

{'boosting_type': 'gbdt',
 'class_weight': None,
 'colsample_bytree': 1.0,
 'importance_type': 'split',
 'learning_rate': 0.1,
 'max_depth': -1,
 'min_child_samples': 20,
 'min_child_weight': 0.001,
 'min_split_gain': 0.0,
 'n_estimators': 100,
 'n_jobs': -1,
 'num_leaves': 31,
 'objective': 'binary',
 'random_state': None,
 'reg_alpha': 0.0,
 'reg_lambda': 0.0,
 'silent': True,
 'subsample': 1.0,
 'subsample_for_bin': 200000,
 'subsample_freq': 0}

In [61]:
from sklearn.model_selection import GridSearchCV

# 試行するパラメータを羅列
params = {
    'num_leaves': [3, 4, 5, 6, 7, 8, 9, 10],
    'reg_alpha': [0, 1, 2, 3, 4, 5,10, 100],
    'reg_lambda': [10, 15, 18, 20, 21, 22, 23, 25, 27, 29]
}

grid_search = GridSearchCV(gbm, param_grid=params, cv=3)
grid_search.fit(X_train, y_train)

print(grid_search.best_score_)
print(grid_search.best_params_)


0.39999999999999997
{'num_leaves': 3, 'reg_alpha': 0, 'reg_lambda': 25}


In [90]:
y_pred_GBM = grid_search.predict(X_test)

In [62]:
print(grid_search.best_estimator_.feature_importances_)

[  0  55  56   0  92 104 123  35  48  35  72   0  61  56  51   0]


In [63]:
importance = pd.DataFrame(grid_search.best_estimator_.feature_importances_, index=x_feature.columns, columns=['importance'])
display(importance.sort_values('importance'))

importance
Number of people sharing the same pot                    0
Number of cultivation field                              0
Sleeping_time                                            0
Gender_replaced                                          0
Total_happness                                          35
Steps                                                   35
Activity_level_2                                        48
Livestock_time                                          51
Age                                                     55
MIL                                                     56
Crop_time                                               56
Economics_time                                          61
percentage_production                                   72
walking min from your house to Wangige market?          92
Agrobiodiversity                                       104
Grasping power(kg)                                     123

In [47]:
params = {
    'num_leaves': [3],
    'reg_alpha': [3],
    'reg_lambda': [22]
}

gbm = lgb.train(
params,
train_data,
valid_sets=eval_data,
num_boost_round=1000,
verbose_eval=4,
)

In [48]:
from sklearn import model_selection

print('== デフォルト・パラメータの場合 ==')
gbm = lgb.LGBMClassifier(objective='binary')

score = model_selection.cross_val_score(gbm, X_train, y_train, cv=3) # cv=3は3分割の意
print('各正解率', score)
print('正解率', score.mean())

print('\n== パラメータをチューニングした場合 ==')
gbm = lgb.LGBMClassifier(objective='binary', num_leaves=3, reg_alpha=3, reg_lambda=22)
score = model_selection.cross_val_score(gbm, X_train, y_train, cv=3) # cv=3は3分割の意
print('各正解率', score)
print('正解率', score.mean())

== デフォルト・パラメータの場合 ==
各正解率 [0.31428571 0.4        0.34285714]
正解率 0.3523809523809524

== パラメータをチューニングした場合 ==
各正解率 [0.34285714 0.37142857 0.42857143]
正解率 0.38095238095238093


Feature Importance:
----

参考：
https://qiita.com/studio_haneya/items/e70e835c26524d506e19

In [49]:
importance = pd.DataFrame(grid_search.best_estimator_.feature_importances_, index=x_feature.columns, columns=['importance'])
display(importance.sort_values('importance'))

importance
Number of people sharing the same pot                    0
Number of cultivation field                              0
Sleeping_time                                            0
Gender_replaced                                          0
Total_happness                                          35
Steps                                                   35
Activity_level_2                                        48
Livestock_time                                          51
Age                                                     55
MIL                                                     56
Crop_time                                               56
Economics_time                                          61
percentage_production                                   72
walking min from your house to Wangige market?          92
Agrobiodiversity                                       104
Grasping power(kg)                                     123

SVMを用いて分析
----

In [50]:
#SVMによるサーチ
from sklearn import svm, metrics

params = [
    {"C": [1,10,100,1000], "kernel":["linear"]},
    {"C": [1,10,100,1000], "kernel":["rbf"], "gamma":[0.001, 0.0001]}
]

# グリッドサーチを行う
clf = GridSearchCV( svm.SVC(), params, n_jobs=-1,cv=3,iid=True )
clf.fit(X_train, y_train)
print("学習モデル=", clf.best_estimator_)

学習モデル= SVC(C=1, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma=0.001, kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)


/usr/local/lib/python3.7/dist-packages/sklearn/model_selection/_search.py:823: FutureWarning: The parameter 'iid' is deprecated in 0.22 and will be removed in 0.24.
  "removed in 0.24.", FutureWarning


In [51]:
type(y_test)
print(y_test)
print(y_test.shape)
y_test = np.array(y_test)
y_test = y_test.reshape(-1,1)

111    4
86     4
81     4
1      2
120    4
0      3
71     3
40     2
3      4
6      4
27     3
128    3
22     2
125    4
115    3
52     4
10     2
126    2
130    2
105    3
77     3
23     2
89     4
11     2
18     2
4      2
91     3
Name: BMI_EVENT, dtype: int64
(27,)


In [52]:
# 検証用データで精度を確認
pre = clf.predict(X_test)
ac_score = metrics.accuracy_score(pre, y_test)
print("正解率=",ac_score)

正解率= 0.3333333333333333


In [54]:
importance = pd.DataFrame(clf.best_estimator_.feature_importances_, index=x_feature.columns, columns=['importance'])
display(importance.sort_values('importance'))

AttributeError: ignored

ニューラルネットワークによる分析：
---

参考URL:
https://teratail.com/questions/234326

In [55]:
from sklearn import preprocessing
from sklearn.neural_network import MLPClassifier

In [56]:
param_grid = {'hidden_layer_sizes': [(100,100),(10,10),(100,),(10,)],
              'max_iter': [10000,1000,100,10]}

grid_search_NN = GridSearchCV(MLPClassifier(), param_grid, cv=3)
grid_search_NN.fit(X_train, y_train)

print(grid_search_NN.best_estimator_)
print(grid_search_NN.best_params_)

/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (10) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (10) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (10) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (100) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)
/us

MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
              beta_2=0.999, early_stopping=False, epsilon=1e-08,
              hidden_layer_sizes=(100, 100), learning_rate='constant',
              learning_rate_init=0.001, max_fun=15000, max_iter=10000,
              momentum=0.9, n_iter_no_change=10, nesterovs_momentum=True,
              power_t=0.5, random_state=None, shuffle=True, solver='adam',
              tol=0.0001, validation_fraction=0.1, verbose=False,
              warm_start=False)
{'hidden_layer_sizes': (100, 100), 'max_iter': 10000}


/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (10) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


In [57]:
# 検証用データで精度を確認
pre_NN = grid_search_NN.predict(X_test)
ac_score = metrics.accuracy_score(pre_NN, y_test)
print("正解率=",ac_score)

正解率= 0.4444444444444444


In [58]:
importance = pd.DataFrame(grid_search_NN.best_estimator_.feature_importances_, index=x_feature.columns, columns=['importance'])
display(importance.sort_values('importance'))

AttributeError: ignored

重回帰分析のグリッドサーチ：
----

In [59]:
from sklearn.linear_model import LogisticRegression
# グリッドサーチを行うためのパラメーター（LogisticRegressionのパラメーター）
parameters = [
    {'solver': ['liblinear', 'saga'], 'penalty':['l1', 'l2'], 'C': [0.1, 1, 10, 100]},
    {'solver': ['newton-cg', 'sag', 'lbfgs' ], 'penalty':['l2'], 'C': [0.1, 1, 10, 100]},
]

#グリッドサーチ実行
classifier = GridSearchCV(LogisticRegression(), parameters, cv=3, n_jobs=-1)
classifier.fit(X_train, y_train)
print("Accuracy score (train): ", classifier.score(X_train, y_train))
print("Accuracy score (test): ", classifier.score(X_test, y_test))
print(classifier.best_estimator_) # ベストのパラメーターを持つ分類器

print(classifier.best_score_)
print(classifier.best_params_)

Accuracy score (train):  0.3904761904761905
Accuracy score (test):  0.4074074074074074
LogisticRegression(C=0.1, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='sag', tol=0.0001, verbose=0,
                   warm_start=False)
0.39999999999999997
{'C': 0.1, 'penalty': 'l2', 'solver': 'sag'}


/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)


In [64]:
# 検証用データで精度を確認
pre_lr = classifier.predict(X_test)
ac_score_ir = metrics.accuracy_score(pre_lr, y_test)
print("正解率=",ac_score_ir)

正解率= 0.4074074074074074


In [65]:
importance = pd.DataFrame(classifier.best_estimator_.feature_importances_, index=x_feature.columns, columns=['importance'])
display(importance.sort_values('importance'))

AttributeError: ignored

ランダムフォレストによるグリッドサーチ
----
参考URL：　https://watlab-blog.com/2020/02/20/grid-search-decisiontree/

In [66]:
from sklearn import tree
# グリッドサーチ用のパラメータを辞書型で設定
param = {'max_depth':[1, 2, 3, 4, 5],
         'min_samples_leaf':[1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
         'min_samples_split':[2, 3, 4, 5]}
 
# 決定木による学習
tree = GridSearchCV(tree.DecisionTreeClassifier(),   # グリッドサーチで決定木を定義
                   param, cv=5, iid=True)
tree.fit(X_train, y_train)                           # フィッティング
 
# スコアとパラメータの組み合わせ
scores = tree.cv_results_['mean_test_score']
params = tree.cv_results_['params']
 
# 結果の確認
best_tree = tree.best_estimator_
print('最良条件:\n', best_tree)
print('訓練スコア:\n', best_tree.score(X_train, y_train))
print('テストスコア:\n', best_tree.score(X_test, y_test))
for i in range(len(scores)):
    print(scores[i], params[i])

最良条件:
 DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=1, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')
訓練スコア:
 0.44761904761904764
テストスコア:
 0.4444444444444444
0.38095238095238093 {'max_depth': 1, 'min_samples_leaf': 1, 'min_samples_split': 2}
0.38095238095238093 {'max_depth': 1, 'min_samples_leaf': 1, 'min_samples_split': 3}
0.38095238095238093 {'max_depth': 1, 'min_samples_leaf': 1, 'min_samples_split': 4}
0.38095238095238093 {'max_depth': 1, 'min_samples_leaf': 1, 'min_samples_split': 5}
0.38095238095238093 {'max_depth': 1, 'min_samples_leaf': 2, 'min_samples_split': 2}
0.38095238095238093 {'max_depth': 1, 'min_samples_leaf': 2, 'min_samples_split': 3}
0.3809523

/usr/local/lib/python3.7/dist-packages/sklearn/model_selection/_search.py:823: FutureWarning: The parameter 'iid' is deprecated in 0.22 and will be removed in 0.24.
  "removed in 0.24.", FutureWarning


In [95]:
y_pred_RR = tree.predict(X_test)

In [78]:
print(best_tree.best_score_)
print(best_tree.best_params_)

AttributeError: ignored

In [67]:
importance = pd.DataFrame(tree.best_estimator_.feature_importances_, index=x_feature.columns, columns=['importance'])
display(importance.sort_values('importance'))

importance
Number of people sharing the same pot                  0.0
Age                                                    0.0
Number of cultivation field                            0.0
walking min from your house to Wangige market?         0.0
Agrobiodiversity                                       0.0
Grasping power(kg)                                     0.0
Total_happness                                         0.0
Activity_level_2                                       0.0
Steps                                                  0.0
percentage_production                                  0.0
Sleeping_time                                          0.0
Economics_time                                         0.0
Crop_time                                              0.0
Livestock_time                                         0.0
Gender_replaced                                        0.0
MIL                                                    1.0

XGboostのによるグリッドサーチ：
---

In [80]:
import xgboost as xgb
from sklearn.model_selection import StratifiedKFold


cv_params = {'objective':['multi:softprob'],  
             'num_class': [4],
             'n_estimators':[500],
             'booster': ['gbtree'],
             'learning_rate':[0.01, 0.03, 0.1, 0.3, 0.5],
             'min_split_loss':[1],
             'max_depth':[1, 3, 7, 8, 9, 10],
             'min_child_weight':[1, 2, 3, 4, 5],
             'subsample':[0.5, 0.75, 1.0],
             'colsample_bytree':[0.5, 0.75, 1.0]
            }

skf = StratifiedKFold(n_splits=4, shuffle=True, random_state=42) 

bst_grid = GridSearchCV(xgb.XGBClassifier(),                           # 推定器（Estimator）
                        param_grid=cv_params,          # パラメータ
                        cv=skf,                        # 交差検証ジェネレーター
                        scoring="accuracy",        # 検証用データの予測を評価する指標
                        n_jobs=1,                      # 並行して実行するジョブの数
                        verbose=10)                     # 学習の経過の表示(非表示)

bst_grid = bst_grid.fit(X_train,                       # 使用するデータセット
                        y_train,
                        early_stopping_rounds=10,      # アーリーストッピング
                        eval_set=[(X_test, y_test)],   # 学習に用いる検証用データ
                        eval_metric='mlogloss',        # アーリーストッピングに使用する指標
                        verbose=100)                     # 学習の経過の表示(非表示)

Fitting 4 folds for each of 1350 candidates, totalling 5400 fits
[CV] booster=gbtree, colsample_bytree=0.5, learning_rate=0.01, max_depth=1, min_child_weight=1, min_split_loss=1, n_estimators=500, num_class=4, objective=multi:softprob, subsample=0.5 
[0]	validation_0-mlogloss:1.38404
Will train until validation_0-mlogloss hasn't improved in 10 rounds.
[100]	validation_0-mlogloss:1.26074
[200]	validation_0-mlogloss:1.20521
Stopping. Best iteration:
[230]	validation_0-mlogloss:1.19802

[CV]  booster=gbtree, colsample_bytree=0.5, learning_rate=0.01, max_depth=1, min_child_weight=1, min_split_loss=1, n_estimators=500, num_class=4, objective=multi:softprob, subsample=0.5, score=0.259, total=   0.2s
[CV] booster=gbtree, colsample_bytree=0.5, learning_rate=0.01, max_depth=1, min_child_weight=1, min_split_loss=1, n_estimators=500, num_class=4, objective=multi:softprob, subsample=0.5 
[0]	validation_0-mlogloss:1.38334
Will train until validation_0-mlogloss hasn't improved in 10 rounds.


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
/usr/local/lib/python3.7/dist-packages/sklearn/preprocessing/_label.py:268: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.2s remaining:    0.0s


[100]	validation_0-mlogloss:1.26038
[200]	validation_0-mlogloss:1.20394
[300]	validation_0-mlogloss:1.17357
Stopping. Best iteration:
[339]	validation_0-mlogloss:1.16532

[CV]  booster=gbtree, colsample_bytree=0.5, learning_rate=0.01, max_depth=1, min_child_weight=1, min_split_loss=1, n_estimators=500, num_class=4, objective=multi:softprob, subsample=0.5, score=0.346, total=   0.2s
[CV] booster=gbtree, colsample_bytree=0.5, learning_rate=0.01, max_depth=1, min_child_weight=1, min_split_loss=1, n_estimators=500, num_class=4, objective=multi:softprob, subsample=0.5 
[0]	validation_0-mlogloss:1.38498
Will train until validation_0-mlogloss hasn't improved in 10 rounds.
[100]	validation_0-mlogloss:1.29916
Stopping. Best iteration:
[155]	validation_0-mlogloss:1.27709

[CV]  booster=gbtree, colsample_bytree=0.5, learning_rate=0.01, max_depth=1, min_child_weight=1, min_split_loss=1, n_estimators=500, num_class=4, objective=multi:softprob, subsample=0.5, score=0.154, total=   0.1s
[CV] booster=

[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    0.4s remaining:    0.0s
[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed:    0.5s remaining:    0.0s


[200]	validation_0-mlogloss:1.18219
[300]	validation_0-mlogloss:1.14411
Stopping. Best iteration:
[384]	validation_0-mlogloss:1.12272

[CV]  booster=gbtree, colsample_bytree=0.5, learning_rate=0.01, max_depth=1, min_child_weight=1, min_split_loss=1, n_estimators=500, num_class=4, objective=multi:softprob, subsample=0.5, score=0.385, total=   0.2s
[CV] booster=gbtree, colsample_bytree=0.5, learning_rate=0.01, max_depth=1, min_child_weight=1, min_split_loss=1, n_estimators=500, num_class=4, objective=multi:softprob, subsample=0.75 
[0]	validation_0-mlogloss:1.38422
Will train until validation_0-mlogloss hasn't improved in 10 rounds.
[100]	validation_0-mlogloss:1.25537
[200]	validation_0-mlogloss:1.19513
[300]	validation_0-mlogloss:1.16819


[Parallel(n_jobs=1)]: Done   4 out of   4 | elapsed:    0.7s remaining:    0.0s


Stopping. Best iteration:
[339]	validation_0-mlogloss:1.15881

[CV]  booster=gbtree, colsample_bytree=0.5, learning_rate=0.01, max_depth=1, min_child_weight=1, min_split_loss=1, n_estimators=500, num_class=4, objective=multi:softprob, subsample=0.75, score=0.296, total=   0.2s
[CV] booster=gbtree, colsample_bytree=0.5, learning_rate=0.01, max_depth=1, min_child_weight=1, min_split_loss=1, n_estimators=500, num_class=4, objective=multi:softprob, subsample=0.75 
[0]	validation_0-mlogloss:1.38394
Will train until validation_0-mlogloss hasn't improved in 10 rounds.
[100]	validation_0-mlogloss:1.26706
[200]	validation_0-mlogloss:1.20803
[300]	validation_0-mlogloss:1.17721


[Parallel(n_jobs=1)]: Done   5 out of   5 | elapsed:    1.0s remaining:    0.0s


Stopping. Best iteration:
[332]	validation_0-mlogloss:1.16805

[CV]  booster=gbtree, colsample_bytree=0.5, learning_rate=0.01, max_depth=1, min_child_weight=1, min_split_loss=1, n_estimators=500, num_class=4, objective=multi:softprob, subsample=0.75, score=0.385, total=   0.2s
[CV] booster=gbtree, colsample_bytree=0.5, learning_rate=0.01, max_depth=1, min_child_weight=1, min_split_loss=1, n_estimators=500, num_class=4, objective=multi:softprob, subsample=0.75 
[0]	validation_0-mlogloss:1.38484
Will train until validation_0-mlogloss hasn't improved in 10 rounds.
[100]	validation_0-mlogloss:1.287
Stopping. Best iteration:
[186]	validation_0-mlogloss:1.26086

[CV]  booster=gbtree, colsample_bytree=0.5, learning_rate=0.01, max_depth=1, min_child_weight=1, min_split_loss=1, n_estimators=500, num_class=4, objective=multi:softprob, subsample=0.75, score=0.192, total=   0.1s
[CV] booster=gbtree, colsample_bytree=0.5, learning_rate=0.01, max_depth=1, min_child_weight=1, min_split_loss=1, n_esti

[Parallel(n_jobs=1)]: Done   6 out of   6 | elapsed:    1.2s remaining:    0.0s
[Parallel(n_jobs=1)]: Done   7 out of   7 | elapsed:    1.3s remaining:    0.0s


[100]	validation_0-mlogloss:1.23723
[200]	validation_0-mlogloss:1.16281
[300]	validation_0-mlogloss:1.1288
[400]	validation_0-mlogloss:1.10617
Stopping. Best iteration:
[429]	validation_0-mlogloss:1.09926

[CV]  booster=gbtree, colsample_bytree=0.5, learning_rate=0.01, max_depth=1, min_child_weight=1, min_split_loss=1, n_estimators=500, num_class=4, objective=multi:softprob, subsample=0.75, score=0.385, total=   0.3s
[CV] booster=gbtree, colsample_bytree=0.5, learning_rate=0.01, max_depth=1, min_child_weight=1, min_split_loss=1, n_estimators=500, num_class=4, objective=multi:softprob, subsample=1.0 
[0]	validation_0-mlogloss:1.38438
Will train until validation_0-mlogloss hasn't improved in 10 rounds.
[100]	validation_0-mlogloss:1.26322
[200]	validation_0-mlogloss:1.20235
[300]	validation_0-mlogloss:1.16477


[Parallel(n_jobs=1)]: Done   8 out of   8 | elapsed:    1.6s remaining:    0.0s


[400]	validation_0-mlogloss:1.14421
[499]	validation_0-mlogloss:1.13203
[CV]  booster=gbtree, colsample_bytree=0.5, learning_rate=0.01, max_depth=1, min_child_weight=1, min_split_loss=1, n_estimators=500, num_class=4, objective=multi:softprob, subsample=1.0, score=0.370, total=   0.3s
[CV] booster=gbtree, colsample_bytree=0.5, learning_rate=0.01, max_depth=1, min_child_weight=1, min_split_loss=1, n_estimators=500, num_class=4, objective=multi:softprob, subsample=1.0 
[0]	validation_0-mlogloss:1.38514
Will train until validation_0-mlogloss hasn't improved in 10 rounds.
[100]	validation_0-mlogloss:1.27238
[200]	validation_0-mlogloss:1.21737
[300]	validation_0-mlogloss:1.18721


[Parallel(n_jobs=1)]: Done   9 out of   9 | elapsed:    1.9s remaining:    0.0s


Streaming output truncated to the last 5000 lines.
Stopping. Best iteration:
[2]	validation_0-mlogloss:1.27565

[CV]  booster=gbtree, colsample_bytree=1.0, learning_rate=0.3, max_depth=1, min_child_weight=1, min_split_loss=1, n_estimators=500, num_class=4, objective=multi:softprob, subsample=0.75, score=0.231, total=   0.0s
[CV] booster=gbtree, colsample_bytree=1.0, learning_rate=0.3, max_depth=1, min_child_weight=1, min_split_loss=1, n_estimators=500, num_class=4, objective=multi:softprob, subsample=0.75 
[0]	validation_0-mlogloss:1.30767
Will train until validation_0-mlogloss hasn't improved in 10 rounds.
Stopping. Best iteration:
[13]	validation_0-mlogloss:1.09214

[CV]  booster=gbtree, colsample_bytree=1.0, learning_rate=0.3, max_depth=1, min_child_weight=1, min_split_loss=1, n_estimators=500, num_class=4, objective=multi:softprob, subsample=0.75, score=0.423, total=   0.0s
[CV] booster=gbtree, colsample_bytree=1.0, learning_rate=0.3, max_depth=1, min_child_weight=1, min_split_loss

[Parallel(n_jobs=1)]: Done 5400 out of 5400 | elapsed: 10.7min finished


Prediction
----

In [81]:
y_pred_xgb = bst_grid.predict(X_test)

In [82]:
print(bst_grid.best_score_)
print(bst_grid.best_params_)

0.40028490028490027
{'booster': 'gbtree', 'colsample_bytree': 1.0, 'learning_rate': 0.5, 'max_depth': 7, 'min_child_weight': 2, 'min_split_loss': 1, 'n_estimators': 500, 'num_class': 4, 'objective': 'multi:softprob', 'subsample': 0.75}


In [83]:
importance = pd.DataFrame(bst_grid.best_estimator_.feature_importances_, index=x_feature.columns, columns=['importance'])
display(importance.sort_values('importance'))

importance
Number of cultivation field                       0.000000
Gender_replaced                                   0.000000
Sleeping_time                                     0.051391
Livestock_time                                    0.056830
Crop_time                                         0.064883
Grasping power(kg)                                0.064909
walking min from your house to Wangige market?    0.067701
Age                                               0.070348
Steps                                             0.071431
Activity_level_2                                  0.074009
Agrobiodiversity                                  0.074984
percentage_production                             0.076561
Total_happness                                    0.077686
MIL                                               0.080625
Number of people sharing the same pot             0.082380
Economics_time                                    0.086263

In [84]:
print("Accuracy score (train): ", bst_grid.score(X_train, y_train))
print('Accuracy:',accuracy_score(y_test,y_pred_xgb))
# print("Accuracy score (test): ", bst_grid.score(X_test, y_test))
print(bst_grid.best_estimator_) # ベストのパラメーターを持つ分類器

Accuracy score (train):  0.9238095238095239
Accuracy: 0.48148148148148145
XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1.0, gamma=0,
              learning_rate=0.5, max_delta_step=0, max_depth=7,
              min_child_weight=2, min_split_loss=1, missing=None,
              n_estimators=500, n_jobs=1, nthread=None, num_class=4,
              objective='multi:softprob', random_state=0, reg_alpha=0,
              reg_lambda=1, scale_pos_weight=1, seed=None, silent=None,
              subsample=0.75, verbosity=1)


現時点までの考察：
---
グリッドサーチを行った結果、ツリー系の分析、回帰系の分析を行った。

正確性：
XGboost = 0.40

ランダムフォレスト　＝　0.44

重回帰分析　＝　0.40

LightGBM　＝　0.39

ニューラルネットワーク　＝　0.44

の結果だった。

数値が上がらない理由
そもそも分類データが少ない。データにバラツキがある。

論文を読み直した結果、モデルの値はMAEとRMSEによって評価していたため、こちらでもその評価を行う。

参考URL: https://funatsu-lab.github.io/open-course-ware/basic-theory/accuracy-index/


In [86]:
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error

In [89]:
#XGBoost
mae_XGB = mean_absolute_error(y_test, y_pred_xgb)
rmse_XGB = np.sqrt(mean_squared_error(y_test, y_pred_xgb))
print(mae_XGB)
print(rmse_XGB)

0.7777777777777778
1.1385500851066221


In [91]:
#LightGBM
mae_GBM = mean_absolute_error(y_test, y_pred_GBM)
rmse_GBM = np.sqrt(mean_squared_error(y_test, y_pred_GBM))
print(mae_GBM)
print(rmse_GBM)

0.9629629629629629
1.247219128924647


In [92]:
#SVM
mae_SVM = mean_absolute_error(y_test, pre)
rmse_SVM = np.sqrt(mean_squared_error(y_test, pre))
print(mae_SVM)
print(rmse_SVM)

1.037037037037037
1.3052600138300812


In [93]:
#ニューラルネットワーク
mae_NN = mean_absolute_error(y_test, pre_NN)
rmse_NN = np.sqrt(mean_squared_error(y_test, pre_NN))
print(mae_NN)
print(rmse_NN)

0.8518518518518519
1.2018504251546631


In [94]:
#重回帰分析
mae_MLR = mean_absolute_error(y_test, pre_lr)
rmse_MLR = np.sqrt(mean_squared_error(y_test, pre_lr))
print(mae_MLR)
print(rmse_MLR)

0.8888888888888888
1.2171612389003692


In [96]:
#ランダムフォレスト
mae_RR = mean_absolute_error(y_test, y_pred_RR)
rmse_RR = np.sqrt(mean_squared_error(y_test, y_pred_RR))
print(mae_RR)
print(rmse_RR)


0.8148148148148148
1.1547005383792515


MAEおよびRMSEの分析の結果：
--
XGboostが最も低いを値を示したが、0.77は依然として高い値である。

今後の方針としては、サンプル数を増やすか、分析する特徴量を減らす必要があるかもしれい。
現在は、説明変数はBMIをカテゴリー変数として扱っている。実数での変化を見るということできる。
また、血圧のイベントについても調査することができる。

オーバーサンプリング：
---
SMOTE

参考URL:
https://toukei-lab.com/imbalance-data-smote

In [97]:
!pip install imblearn

In [98]:
from imblearn.over_sampling import SMOTE
sm = SMOTE()
x_resampled, y_resampled = sm.fit_sample(X_train, y_train)

/usr/local/lib/python3.7/dist-packages/sklearn/externals/six.py:31: FutureWarning: The module is deprecated in version 0.21 and will be removed in version 0.23 since we've dropped support for Python 2.7. Please rely on the official version of six (https://pypi.org/project/six/).
  "(https://pypi.org/project/six/).", FutureWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:144: FutureWarning: The sklearn.neighbors.base module is  deprecated in version 0.22 and will be removed in version 0.24. The corresponding classes / functions should instead be imported from sklearn.neighbors. Anything that cannot be imported from sklearn.neighbors is now part of the private API.
  warnings.warn(message, FutureWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function safe_indexing is deprecated; safe_indexing is deprecated in version 0.22 and will be removed in version 0.24.
  warnings.warn(msg, category=FutureWarning)
/usr/loca

In [105]:
print(y_resampled.shape)
print('=======================')
print(np.array(y_train.shape))

(156,)
[105]


SMOTE後のLightGBM：
---

In [115]:
# 試行するパラメータを羅列
params = {
    'num_leaves': [3, 4, 5, 6, 7, 8, 9, 10],
    'reg_alpha': [0, 1, 2, 3, 4, 5,10, 100],
    'reg_lambda': [10, 15, 18, 20, 21, 22, 23, 25, 27, 29]
}

grid_search_SMOTE = GridSearchCV(gbm, param_grid=params, cv=3)
grid_search_SMOTE.fit(x_resampled, y_resampled)

print(grid_search_SMOTE.best_score_)
print(grid_search_SMOTE.best_params_)

0.5769230769230769
{'num_leaves': 5, 'reg_alpha': 0, 'reg_lambda': 10}


In [116]:
#予測XGBOOST
y_pred_GBM_SMOTE = grid_search_SMOTE.predict(X_test)

In [117]:
importance = pd.DataFrame(grid_search_SMOTE.best_estimator_.feature_importances_, index=x_feature.columns, columns=['importance'])
display(importance.sort_values('importance'))

importance
Number of cultivation field                             13
Number of people sharing the same pot                   43
Gender_replaced                                         53
Livestock_time                                          60
Steps                                                   74
walking min from your house to Wangige market?          88
Economics_time                                          96
Crop_time                                              104
Sleeping_time                                          106
Total_happness                                         107
Activity_level_2                                       112
MIL                                                    128
Agrobiodiversity                                       129
Grasping power(kg)                                     129
percentage_production                                  156
Age                                                    176

In [118]:
#XGBoost_SMOTE
mae_XGB_SMOTE = mean_absolute_error(y_test, y_pred_GBM_SMOTE)
rmse_XGB_SMOTE = np.sqrt(mean_squared_error(y_test, y_pred_GBM_SMOTE))
print(mae_XGB_SMOTE)
print(rmse_XGB_SMOTE)

0.9629629629629629
1.247219128924647


SMOTEしたランダムフォレスト：
---

In [120]:
from sklearn import tree
# グリッドサーチ用のパラメータを辞書型で設定
param = {'max_depth':[1, 2, 3, 4, 5],
         'min_samples_leaf':[1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
         'min_samples_split':[2, 3, 4, 5]}
 
# 決定木による学習
tree_SMOTE = GridSearchCV(tree.DecisionTreeClassifier(),   # グリッドサーチで決定木を定義
                   param, cv=5, iid=True)
tree_SMOTE.fit(x_resampled, y_resampled)                           # フィッティング
 
# スコアとパラメータの組み合わせ
scores = tree_SMOTE.cv_results_['mean_test_score']
params = tree_SMOTE.cv_results_['params']
 
# 結果の確認
best_tree = tree_SMOTE.best_estimator_
print('最良条件:\n', best_tree)
print('訓練スコア:\n', best_tree.score(X_train, y_train))
print('テストスコア:\n', best_tree.score(X_test, y_test))

最良条件:
 DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=5, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=2, min_samples_split=4,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')
訓練スコア:
 0.6952380952380952
テストスコア:
 0.4074074074074074


/usr/local/lib/python3.7/dist-packages/sklearn/model_selection/_search.py:823: FutureWarning: The parameter 'iid' is deprecated in 0.22 and will be removed in 0.24.
  "removed in 0.24.", FutureWarning


In [125]:
y_pred_RR_SMOTE = best_tree.predict(X_test)

In [126]:
#RR_SMOTE
mae_RR_SMOTE = mean_absolute_error(y_test, y_pred_RR_SMOTE)
rmse_RR_SMOTE = np.sqrt(mean_squared_error(y_test, y_pred_RR_SMOTE))
print(mae_RR_SMOTE)
print(rmse_RR_SMOTE)

0.8148148148148148
1.1547005383792515
